ctmap: GC orphan SNAT entries

brb opened the pull request:
See commit msg.

Fixes #12686

Follow-up:
- [ ] Handle orphan DSR SNAT entries
- [x] Add metrics

brb committed:
ctmap: GC orphan SNAT entries

This commit adds a mechanism to remove orphan SNAT entries.

We call an SNAT entry orphan if it does not have either a corresponding
CT entry or an SNAT entry in a reverse order. Both cases can happen due
to LRU eviction heuristics (both CT and NAT maps are of the LRU type).

The mechanism for the removal is based on the GC signaling in the
datapath. When the datapath SNAT routine fails to find a free mapping
after SNAT_SIGNAL_THRES attempts, it sends the signal via the perf ring
buffer. The consumer of the buffer is the daemon. After receiving the
signal it invokes the CT GC.

The newly implemented GC addition iterates over all SNAT entries and
checks whether a corresponding CT entry is found, and if not, it tries
to remove both SNAT entries (for original and reverse flows).

For now, I didn't add GC of orphan SNAT entries created by DSR to keep
complexity of changes as low as possible. This will come as a follow up.

Signed-off-by: Martynas Pumputis <m@lambda.lt>

borkmann commented on pkg/maps/ctmap/ctmap.go:
```diff
@@ -510,6 +517,102 @@ func GC(m *Map, filter *GCFilter) int {
+			ctKey.DestPort = natKey.DestPort
+			ctKey.NextHeader = natKey.NextHeader
+			// TODO(brb) Check for CT entry regardless of direction
+			if _, err := ctMap.Lookup(&ctKey); err != nil && errors.Is(err, unix.ENOENT) {
```
I don't quite follow why we do this SNAT map GC here in addition to the regular one? That seems a _very_ high CPU cost to pay. When we get the signal from BPF data path that there is pressure in finding a new SNAT mapping in the NAT table then idea was to walk all NAT entries, find orphans but given you already do the CT lookup here anyway, we already have the related CT entry at hand too, so it's low-cost to delete it from right here. That means instead of walking CT + SNAT map, we just swap walking CT map w/ NAT map and only delete CT from there. Is there any rationale that this is not feasible? Is the concern that we also want to purge CT_SERVICE ones?

brb replied:
> Is there any rationale that this is not feasible?

Maybe I'm missing something, but I don't see how it's possible, because of CT_SERVICE entries and also, not all CT_{INGRESS,EGRESS} entries have related NAT entries (e.g. pod to pod).

borkmann replied:
Right, I mean technically, they would be evicted via LRU or regular time-based GC. The CT_SERVICE ones are more annoying though given we want to improve triggering new backend selection. Technically it's an entirely different issue and cleanups could be triggered on backend updates, but the regular CT walk as you have in addition in here would help it too, just the cost becomes huge.

joestringer commented on pkg/maps/ctmap/gc/gc.go:
```diff
@@ -196,6 +199,17 @@ func runGC(e *endpoint.Endpoint, ipv4, ipv6 bool, filter *ctmap.GCFilter) (mapTy
+
+		if triggeredBySignal {
+			stats := ctmap.PurgeOrphanNATEntries(m)
+			if stats != nil && (stats.EgressDeleted != 0 || stats.IngressDeleted != 0) {
```
Did you consider adding metrics so that production users would have a way to introspect these? I feel like it could potentially be useful for debugging since there would otherwise be no way to get these stats (assuming debug logs in production is not possible, which I would argue given how noisy they are).

brb replied:
Yeah, we definitely need the metric for it. I will add it as a follow up.

joestringer commented on pkg/maps/ctmap/ctmap.go:
```diff
@@ -510,6 +517,102 @@ func GC(m *Map, filter *GCFilter) int {
+
+		if natKey.GetFlags()&tuple.TUPLE_F_IN == 0 { // dir == CT_EGRESS
+			ctKey.Flags = tuple.TUPLE_F_OUT
+			// Workaround #5848
```
Kinda surprised we don't have a `copyKeyWorkingAround5848()` function :sweat_smile:

brb replied:
:laughing:

joestringer commented on pkg/maps/ctmap/ctmap.go:
```diff
@@ -510,6 +517,108 @@ func GC(m *Map, filter *GCFilter) int {
+				NextHeader: natKey.NextHeader,
+				Flags:      tuple.TUPLE_F_OUT,
+			}}
+			if _, err := ctMap.Lookup(ctKey); err != nil && errors.Is(err, unix.ENOENT) {
```
I'm curious, do we need the != nil check?
```suggestion
			if _, err := ctMap.Lookup(ctKey); errors.Is(err, unix.ENOENT) {
```

joestringer replied:
(Completely understandable if we're trying to avoid reflection in `errors.Is()` or something)

joestringer commented on pkg/maps/ctmap/ctmap.go:
```diff
@@ -510,6 +517,108 @@ func GC(m *Map, filter *GCFilter) int {
+		natKey := key.(*nat.NatKey4)
+		natVal := value.(*nat.NatEntry4)
+
+		if natKey.GetFlags()&tuple.TUPLE_F_IN == 1 { // natKey is r(everse)tuple
```
I assume this is because we only apply e.g. SNAT on nodeport on the way back out of the intermediate node, or on the way out from the local endpoint/application... hence the `TUPLE_F_IN` would only be set for reply tuples. :+1:

joestringer commented on pkg/maps/ctmap/ctmap.go:
```diff
@@ -510,6 +517,108 @@ func GC(m *Map, filter *GCFilter) int {
+			// Note, that in the datapath otuple is inserted into the map after
+			// rtuple. This means that if we check first for rtuple and remove
+			// otuple, then there would be a race window during which to-be
+			// inserted NAT tuples by the datapath could be destroyed by the GC.
```
I think this is handling the case:

1. Datapath creates CT entry
1. Datapath creates SNAT reverse tuple entry
1. The CT entry is deleted for some reason (?)
1. Userspace dumps the SNAT reverse tuple entry, decides to delete the SNAT reverse tuple
1. Datapath creates the SNAT forward tuple
1. Now there's an original-dir tuple but no reverse tuple.

For a given new connection, I would expect the datapath is going through steps 1,2,5 in quick succession, so I don't understand how step 3 gets interleaved. It's a new connection, so presumably the CT entry won't be timed out & GC'd. Unless you're cycling the entire CT map within the time of handling a single packet, LRU wouldn't evict the new entry.

If the above is the case, it seems like this entire alternate conditional branch is unnecessary optimization.

Is there another case you have in mind for this branch (handling `!TUPLE_F_IN`)?

joestringer replied:
(from discussion with @jrfastab out-of-band)

brb replied:
Cross-posting @jrfastab question from the backport PR to keep everything in the single place:

> @brb I'm trying to understand the if/else branches here. I understand the flow of CT and SNAT map population as follows,
>
>    Create a CT entry
>   Create SNAT reply entry (TUPLE_F_IN) ->
>    Create SNAT original dir entry (TUPLE_F_OUT)
>
> My understanding is then for most cases we should have a key in both directions. At some point the CT entry gets deleted and we have an orphaned SNAT entry that needs to be deleted.
>
> At this point we start walking our nat table. If the key has TUPLE_F_IN set AND no CT Entry we will delete both oNatKey and rNatKey. All good as far as I read it.
>
> Now if the key has TUPLE_F_OUT set we use the else branch and check if its rNatKey exists. If it does leave it alone. If not its somehow been orphaned by its pair. Couple questions here.
>
>    This case where rNatKey of a TUPLE_F_OUT does not exist should be rare correct?
>  It appears to rely on a race condition. Are there other TUPLE_F_* types where it might be more common?
>
> Now working on my assumption its rare, why bother with the else branch at all? Just check for CT in both  > TUPLE_F_IN and TUPLE_F_OUT cases and if it does not exist delete the pair of keys as done in the if branch. > In the race condition handled now in the else branch it would simply fail the delete, no harm done. And by it > being rare who cares if we rarely do an extra delete on the map.
>
> Did I miss something? Seems cleaner to just handle the single case and not have an if/else here if possible.
>
> And just to be clear, I'm not arguing it is incorrect as is, but do want to be sure I understand it correctly.

brb replied:
To answer to both, I thought about handling the `TUPLE_F_OUT` for cases when `TUPLE_F_IN` entries get evicted by the LRU mechanism. However, in the end I think that it's unnecessary optimization, as the datapath should handle this:
```
static __always_inline int __snat_update(void *map, void *otuple, void *ostate,
					 void *rtuple, void *rstate)
{
	int ret = map_update_elem(map, rtuple, rstate, BPF_NOEXIST);
	if (!ret) {
		ret = map_update_elem(map, otuple, ostate, BPF_NOEXIST);
		if (ret)
			map_delete_elem(map, rtuple);
	}
	return ret;
}
```
I.e. if we end up having a `TUPLE_F_IN` entry w/o a corresponding `TUPLE_F_OUT`, next time a packet matches `TUPLE_F_IN`, the `__snat_update()` will delete it at the cost of dropping the packet. IMO this cost is lower than having more expensive GC.

Thanks for pointing this out, I'm going to remove the if-branch.

brb commented on pkg/maps/ctmap/gc/gc.go:
```diff
@@ -111,9 +112,11 @@ func Enable(ipv4, ipv6 bool, restoredEndpoints []*endpoint.Endpoint, mgr Endpoin
 				signal.MuteChannel(signal.SignalWakeGC)
 			}
 
+			triggeredBySignal = false
```
> Why not also purge the nat table at some regular interval in the 'case <-time.After(...)' below. It seems running from below would also help cleanup any orphaned entries before the problem is hit. Or are we trying to save CPU here?

@jrfastab Yeah, my main concern is the CPU costs. Under normal circumstances, there shouldn't be any orphan SNAT entries, thus walking the SNAT table each time the CT GC is invoked would be wasteful in terms of CPU cycles.

brb commented on pkg/maps/ctmap/ctmap.go:
```diff
@@ -510,6 +517,143 @@ func GC(m *Map, filter *GCFilter) int {
+	return &stats
+}
+
+func oNatKeyFromReverse(k nat.NatKey, v nat.NatEntry) nat.NatKey {
```
I didn't move `oNatKeyFromReverse()` and `egressCTKeyFromIngressNatKeyAndVal` to `pkg/maps/nat` as these functions are not safe to be widely reused. To make them safe, we would need to copy addrs, but that would hinder the NAT GC performance.

joestringer commented on pkg/maps/ctmap/ctmap.go:
```diff
@@ -510,6 +517,143 @@ func GC(m *Map, filter *GCFilter) int {
+		// SNAT map. Therefore, to avoid a case when the given ctMap does not
+		// store entries of the given natKey.NextHeader proto, we should return
+		// early. Otherwise, the natKey entries will be removed, which is wrong.
+		if (natKey.GetNextHeader() == u8proto.TCP) != isCTMapTCP {
```
```suggestion
		if (natKey.GetNextHeader() == u8proto.TCP) != ctMap.isTCP() {
```

(Then there's no need for the `isCTMapTCP` variable either)

joestringer commented on pkg/maps/ctmap/ctmap.go:
```diff
@@ -510,6 +517,143 @@ func GC(m *Map, filter *GCFilter) int {
+		natVal := value.(*nat.NatEntry4)
+
+		// In opposite to the CT maps, TCP and UDP entries are stored in the same
+		// SNAT map. Therefore, to avoid a case when the given ctMap does not
```
nit: 

```suggestion
		// Unlike the CT maps, the SNAT map stores TCP and UDP entries in the same map.
		// Therefore, to avoid a case when the given ctMap does not
```

joestringer commented on pkg/maps/ctmap/ctmap.go:
```diff
@@ -510,6 +517,143 @@ func GC(m *Map, filter *GCFilter) int {
+		// SNAT map. Therefore, to avoid a case when the given ctMap does not
+		// store entries of the given natKey.NextHeader proto, we should return
+		// early. Otherwise, the natKey entries will be removed, which is wrong.
+		if (natKey.GetNextHeader() == u8proto.TCP) != isCTMapTCP {
```
Structurally this seems inefficient, we must iterate the entire NAT map twice to handle lookups for the TCP and non-TCP connections?

brb replied:
Good idea! Let me see how I can plug both CT maps into `runGC()`.

joestringer reviewed (commented):
The latest changes make functional sense, but seem inefficient. Can we do better, pass both the TCP and non-TCP CT maps into a single natmap iterator?

brb committed:
ctmap: Iterate SNAT map once when doing GC

Previously, after receiving the signal from the datapath, we iterated
NAT map twice: first to compare against CT TCP map, second - against CT
any map.

Obviously, doing the iterations two times was inefficient. This commit
fixes that by passing both CT {TCP,any} maps to the NAT GC routine. This
allows the NAT GC to iterate once.

Suggested-by: Joe Stringer <joe@cilium.io>
Signed-off-by: Martynas Pumputis <m@lambda.lt>

joestringer commented on pkg/maps/ctmap/ctmap.go:
```diff
@@ -510,6 +517,108 @@ func GC(m *Map, filter *GCFilter) int {
+			case mapTypeIPv6TCPLocal, mapTypeIPv6TCPGlobal:
+				ctMapTCP = m
+			case mapTypeIPv6AnyLocal, mapTypeIPv6AnyGlobal:
+				ctMapAny = m
```
Logic looks right. We might be able to fold it in with the maptype helpers in `pkg/maps/ctmap/types.go` for better code reuse but it's not such a big deal.
